<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="coiaf">
<title>Simulator details • coiaf</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Lato-0.4.4/font.css" rel="stylesheet">
<link href="../deps/Fira_Code-0.4.4/font.css" rel="stylesheet">
<link href="../deps/Fira_Sans-0.4.4/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulator details">
<meta property="og:description" content="coiaf">
<meta property="og:image" content="https://bailey-lab.github.io/coiaf/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">coiaf</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/example_real_data.html">Real data example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Sensitivity analysis</h6>
    <a class="dropdown-item" href="../articles/sensitivity_analysis_continuous.html">Continuous sensitivity analysis</a>
    <a class="dropdown-item" href="../articles/sensitivity_analysis_discrete.html">Discrete sensitivity analysis</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/index.html">More articles...</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-news">News</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-news">
    <a class="dropdown-item" href="../news/index.html">Change log</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bailey-lab/coiaf/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulator details</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bailey-lab/coiaf/blob/HEAD/vignettes/simulator_details.Rmd" class="external-link"><code>vignettes/simulator_details.Rmd</code></a></small>
      <div class="d-none name"><code>simulator_details.Rmd</code></div>
    </div>

    
    
<p>In this package, synthetic sequencing data are generated using a
statistical model for an individual in a given population. Each
individual is assigned a COI value, which is used to simulate the number
of sequence reads mapped to the reference and alternative allele for the
biallelic SNPs considered. These are then used to derive the
within-sample frequency of the population-level minor allele at each
locus <span class="math inline">\(i\)</span>, denoted as <span class="math inline">\(w_i\)</span>.</p>
<p>We first define the number of loci being simulated, <span class="math inline">\(l\)</span>. We assume that the distribution of
reference allele frequencies for locus <span class="math inline">\(i\)</span>, <span class="math inline">\(R_i\)</span>, is described by a Beta distribution
with shape parameters <span class="math inline">\(\alpha\)</span> and
<span class="math inline">\(\beta\)</span>: <span class="math display">\[
    \mathbb{P}(R_i = r_i) =
\frac{\Gamma(\alpha+\beta)}{\Gamma(\alpha)\Gamma(\beta)}r_i^{(\alpha-1)}(1-r_i)^{(\beta-1)},
\]</span></p>
<p>where <span class="math inline">\(\Gamma\)</span> is the Gamma
function. In our simulations, we assume <span class="math inline">\(\alpha = 1\)</span> and <span class="math inline">\(\beta = 5\)</span> and draw <span class="math inline">\(l\)</span> values of <span class="math inline">\(R_i\)</span>. In our methods, we have defined
relationships with respect to the frequency of the minor allele, <span class="math inline">\(p_i\)</span>. Recall that <span class="math inline">\(p_i \in [0, 0.5]\)</span>. Thus, we define <span class="math display">\[
    p_i = \begin{cases}
          r_i &amp; r_i \leq 0.5 \\
          1 - r_i &amp; r_i &gt; 0.5.
       \end{cases}
\]</span></p>
<p>To simulate <span class="math inline">\(w_i\)</span> for each
individual, we first assign the COI, <span class="math inline">\(k\)</span>. Next, we simulate the genotype at
locus <span class="math inline">\(i\)</span>, drawing <span class="math inline">\(k\)</span> values from a Binomial distribution
with probability <span class="math inline">\(p_i\)</span>, i.e., the
probability of having the minor allele at locus <span class="math inline">\(i\)</span> is equal to the population-level
frequency of the minor allele. This yields the matrix, <span class="math inline">\(\mathbb{G} \in \mathbb{R}^{k \times l}\)</span>,
with <span class="math inline">\(k\)</span> rows and <span class="math inline">\(l\)</span> columns defining the phased haplotype
of each strain, where <span class="math inline">\(\mathbb{G}_{i,j} =
1\)</span> indicates that strain <span class="math inline">\(j\)</span>
at locus <span class="math inline">\(i\)</span> has the minor
allele.</p>
<p>To determine the number of sequence reads related to each strain, we
first draw the proportion of each strain, <span class="math inline">\(\mathbf{S} = (S_1, \ldots, S_k)\)</span>, in an
individual with <span class="math inline">\(k\)</span> strains, which we
assume is described by a Dirichlet distribution with concentration
parameters <span class="math inline">\(\alpha_1, \ldots,
\alpha_k\)</span>. The “true” WSMAF of the minor allele at locus <span class="math inline">\(i\)</span>, denoted <span class="math inline">\(\tau_i\)</span>, is thus given by the sum of the
strain proportions for strains with the minor allele. We can express
this using matrix multiplication: <span class="math display">\[
    \boldsymbol \tau = \mathbf{S} \times \mathbb{G}_i,
\]</span> where <span class="math inline">\(\boldsymbol \tau \in
\mathbb{R}^{1 \times l}\)</span>, <span class="math inline">\(\mathbf{S}
\in \mathbb{R}^{1 \times k}\)</span>, and <span class="math inline">\(\mathbb{G} \in \mathbb{R}^{k \times
l}\)</span>.</p>
<p>In simulations with no assumed sequence error, the number of sequence
reads with the minor allele at each locus is given by sampling from a
Binomial distribution <span class="math inline">\(c\)</span> times with
probability <span class="math inline">\(\boldsymbol \tau\)</span>, where
<span class="math inline">\(c\)</span> is the assumed number of sequence
reads, i.e., the read depth or coverage. However, in simulations with
sequence error, we perturb the “true” WSMAF by assuming that a number of
sequence reads are incorrectly sampled. An incorrectly sampled locus
with the major allele will yield the minor allele and vice versa. In our
simulations, we assume a fixed sequence error, <span class="math inline">\(\epsilon\)</span>, such that the probability of
correctly sampling the minor allele is <span class="math inline">\(1 -
\epsilon\)</span>, and the probability of incorrectly sampling the minor
allele is <span class="math inline">\(\epsilon\)</span>. Therefore, we
can represent the probability of sampling the minor allele, now denoted
as <span class="math inline">\(\boldsymbol{\tau}_\epsilon\)</span> to
indicate the added error, as <span class="math display">\[
    \boldsymbol{\tau}_\epsilon = \boldsymbol \tau(1 - \epsilon) + (1 -
\boldsymbol \tau)\epsilon.
\]</span></p>
<p>Lastly, we may account for overdispersion—additional unexpected
variability in our data—and sample the number of sequence reads with the
minor allele at each locus from a Beta-binomial distribution rather than
a Binomial distribution. Dividing the number of instances of the minor
allele by the coverage at each locus results in the simulated WSMAF.</p>
<div class="section level2">
<h2 id="example-simulation">Example simulation<a class="anchor" aria-label="anchor" href="#example-simulation"></a>
</h2>
<p>Our package defines a function <code><a href="../reference/sim_biallelic.html">sim_biallelic()</a></code> which can
easily be used to simulate data for a given sample. To use it, the user
may simply specify the COI of the simulated sample and the population
level minor allele frequency (PLMAF). The outputs of this function
include the phased haplotypes and the within sample minor allele
frequencies (WSMAF) at each locus.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simulation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_biallelic.html">sim_biallelic</a></span><span class="op">(</span><span class="fl">2</span>, plmaf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">0</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">simulation</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulator_details_files/figure-html/sim%20biallelic-1.png" width="700"></p>
<p>Internally, the function goes through several steps to generate the
final output. The first step in generating simulated data is to generate
the population level minor allele frequency (PLMAF). To do so, we sample
from a Beta distribution.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the seed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define number of loci and the PLMAF</span></span>
<span><span class="va">n_loci</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">plmaf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n_loci</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">plmaf</span><span class="op">[</span><span class="va">plmaf</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">plmaf</span><span class="op">[</span><span class="va">plmaf</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">]</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">Locus 1</th>
<th align="right">Locus 2</th>
<th align="right">Locus 3</th>
<th align="right">Locus 4</th>
<th align="right">Locus 5</th>
</tr></thead>
<tbody><tr class="odd">
<td align="right">0.356</td>
<td align="right">0.13</td>
<td align="right">0.012</td>
<td align="right">0.105</td>
<td align="right">0.435</td>
</tr></tbody>
</table>
<p>In cases in which the COI is greater than 1, we will have multiple
strains of differing proportions in our sample. In order to determine
the proportion of each strain present in our sample, we sample from a
Dirichlet distribution.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">strain_p</span> <span class="op">&lt;-</span> <span class="fu">coiaf</span><span class="fu">:::</span><span class="fu"><a href="../reference/rdirichlet.html">rdirichlet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">alpha</span>, <span class="va">COI</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Proportion</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Strain 1</td>
<td align="right">0.296</td>
</tr>
<tr class="even">
<td align="left">Strain 2</td>
<td align="right">0.379</td>
</tr>
<tr class="odd">
<td align="left">Strain 3</td>
<td align="right">0.324</td>
</tr>
</tbody>
</table>
<p>In order to determine the within sample minor allele frequency
(WSMAF) at each locus, we must determine the phased haplotype of each
strain. To do so, we draw from a Binomial distribution with
probabilities given by the PLMAF.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haplotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">COI</span>, <span class="fl">1</span>, <span class="va">x</span><span class="op">)</span>, x <span class="op">=</span> <span class="va">plmaf</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Locus 1</th>
<th align="right">Locus 2</th>
<th align="right">Locus 3</th>
<th align="right">Locus 4</th>
<th align="right">Locus 5</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Strain 1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">Strain 2</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">Strain 3</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>To determine the true WSMAF, we can simply combine our haplotype
information with our strain proportions—finding the sum of the strain
proportions for strains with the minor allele at a particular locus.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">true_wsmaf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">haplotype</span>, <span class="fl">1</span>, <span class="va">strain_p</span>, <span class="st">"*"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rounding errors can cause numbers greater than 1</span></span>
<span><span class="va">true_wsmaf</span><span class="op">[</span><span class="va">true_wsmaf</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1L</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Locus 1</th>
<th align="right">Locus 2</th>
<th align="right">Locus 3</th>
<th align="right">Locus 4</th>
<th align="right">Locus 5</th>
</tr></thead>
<tbody><tr class="odd">
<td align="left">WSMAF</td>
<td align="right">0.704</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.379</td>
</tr></tbody>
</table>
<p>Lastly, to generate our final simulated WSMAF we account for
sequencing error, sample the number of sequence reads with the minor
allele at each locus, and determine the WSMAF by dividing the number of
reads with the minor allele by the coverage at each locus.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Account for sequencing error</span></span>
<span><span class="va">true_wsmaf</span> <span class="op">&lt;-</span> <span class="va">true_wsmaf</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">epsilon</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">true_wsmaf</span><span class="op">)</span> <span class="op">*</span> <span class="va">epsilon</span></span>
<span></span>
<span><span class="co"># Sample sequence reads</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">n_loci</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">coverage</span>, <span class="va">n_loci</span><span class="op">)</span>, prob <span class="op">=</span> <span class="va">true_wsmaf</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Determine simulated WSMAF</span></span>
<span><span class="va">sim_wsmaf</span> <span class="op">&lt;-</span> <span class="va">counts</span> <span class="op">/</span> <span class="va">coverage</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Locus 1</th>
<th align="right">Locus 2</th>
<th align="right">Locus 3</th>
<th align="right">Locus 4</th>
<th align="right">Locus 5</th>
</tr></thead>
<tbody><tr class="odd">
<td align="left">WSMAF</td>
<td align="right">0.73</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.35</td>
</tr></tbody>
</table>
</div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://www.arispas.com" class="external-link">Aris Paschalidis</a>, <a href="https://www.imperial.ac.uk/people/o.watson15" class="external-link">OJ Watson</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
